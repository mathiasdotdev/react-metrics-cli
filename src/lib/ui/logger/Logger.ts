import chalk from 'chalk';

// Export des niveaux pour compatibilit√©
export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  SUCCESS = 2,
  WARN = 3,
  ERROR = 4,
}

// Logger simple utilisant chalk pour la colorisation
export class Logger {
  private static currentLevel: LogLevel = LogLevel.DEBUG;

  static setLevel(level: LogLevel): void {
    Logger.currentLevel = level;
  }

  static newLine(): void {
    console.log('\n');
  }

  static debug(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.DEBUG) {
      console.log(chalk.gray(`üêõ ${message}`), ...args);
    }
  }

  static log(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.INFO) {
      console.log(message, ...args);
    }
  }

  static info(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.INFO) {
      console.log(`üí° ${message}`, ...args);
    }
  }

  static success(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.SUCCESS) {
      console.log(chalk.green(`‚úÖ ${message}`), ...args);
    }
  }

  static warn(message: string, ..._args: any[]): void {
    if (Logger.currentLevel <= LogLevel.WARN) {
      console.warn(chalk.yellow(`‚ö†Ô∏è ${message}`));
    }
  }

  static error(message: string, ..._args: any[]): void {
    if (Logger.currentLevel <= LogLevel.ERROR) {
      console.error(chalk.red(`‚ùå ${message}`));
    }
  }

  static config(message: string, ...args: any[]): void {
    console.log(`üîê ${message}`, ...args);
  }

  static files(message: string, ...args: any[]): void {
    console.log(`üìÅ ${message}`, ...args);
  }

  static cleanup(message: string, ..._args: any[]): void {
    console.warn(chalk.yellow(`üóëÔ∏è ${message}`));
  }

  static settings(message: string, ...args: any[]): void {
    console.log(`‚öôÔ∏è ${message}`, ...args);
  }

  static analysis(message: string, ...args: any[]): void {
    console.log(`üîç ${message}`, ...args);
  }

  static examples(message: string, ...args: any[]): void {
    console.log(`‚úçÔ∏è ${message}`, ...args);
  }

  static report(message: string, ...args: any[]): void {
    console.log(`üìä ${message}`, ...args);
  }

  static list(message: string, ...args: any[]): void {
    console.log(`- ${message}`, ...args);
  }

  // M√©thodes pour l'affichage avec couleurs (terminal output)
  static section(message: string, ...args: any[]): void {
    console.log(chalk.blue(message), ...args);
  }

  static filePath(message: string, ...args: any[]): void {
    console.log(chalk.green(message), ...args);
  }

  static label(message: string, ...args: any[]): void {
    console.log(chalk.gray(message), ...args);
  }

  // M√©thode pour les messages avec couleur personnalis√©e
  static colored(color: string, message: string, ...args: any[]): void {
    console.log(message, ...args);
  }
}
