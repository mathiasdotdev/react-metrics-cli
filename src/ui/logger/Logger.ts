import chalk from 'chalk'

export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  SUCCESS = 2,
  WARN = 3,
  ERROR = 4,
}

export class Logger {
  private static currentLevel: LogLevel = LogLevel.INFO

  static setLevel(level: LogLevel): void {
    Logger.currentLevel = level
  }

  static newLine(): void {
    console.log('\n')
  }

  static debug(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.DEBUG) {
      console.log(chalk.gray('🐛'), message, ...args)
    }
  }

  static log(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.INFO) {
      console.log(chalk.gray(message), ...args)
    }
  }

  static info(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.INFO) {
      console.log(chalk.blue('💡'), message, ...args)
    }
  }

  static success(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.SUCCESS) {
      console.log(chalk.green('✅'), message, ...args)
    }
  }

  static warn(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.WARN) {
      console.log(chalk.yellow('⚠️'), message, ...args)
    }
  }

  static error(message: string, ...args: any[]): void {
    if (Logger.currentLevel <= LogLevel.ERROR) {
      console.error(chalk.red('❌'), message, ...args)
    }
  }

  // Méthodes spécialisées avec icônes
  static credentials(message: string, ...args: any[]): void {
    console.log(chalk.blue('🔑'), message, ...args)
  }

  static download(message: string, ...args: any[]): void {
    console.log(chalk.blue('⬇️'), message, ...args)
  }

  static config(message: string, ...args: any[]): void {
    console.log(chalk.cyan('🔐'), message, ...args)
  }

  static files(message: string, ...args: any[]): void {
    console.log(chalk.green('📁'), message, ...args)
  }

  static cleanup(message: string, ...args: any[]): void {
    console.log(chalk.yellow('🗑️'), message, ...args)
  }

  static settings(message: string, ...args: any[]): void {
    console.log(chalk.blue('⚙️'), message, ...args)
  }

  static analysis(message: string, ...args: any[]): void {
    console.log(chalk.cyan('🔍'), message, ...args)
  }

  static examples(message: string, ...args: any[]): void {
    console.log(chalk.cyan('✍️'), message, ...args)
  }

  static report(message: string, ...args: any[]): void {
    console.log(chalk.cyan('📊'), message, ...args)
  }

  static list(message: string, ...args: any[]): void {
    console.log('\t' + chalk.gray('➖'), message, ...args)
  }

  // Méthode pour les messages avec couleur personnalisée
  static colored(
    color: keyof typeof chalk,
    message: string,
    ...args: any[]
  ): void {
    console.log((chalk as any)[color](message), ...args)
  }
}
