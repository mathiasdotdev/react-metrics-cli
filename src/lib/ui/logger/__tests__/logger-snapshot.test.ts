import { afterEach, beforeEach, describe, expect, test, spyOn } from 'bun:test';
import chalk from 'chalk';
import { Logger, LogLevel } from '../Logger';

describe('Logger - Snapshot Tests', () => {
  let consoleLogSpy: ReturnType<typeof spyOn>;
  let consoleWarnSpy: ReturnType<typeof spyOn>;
  let consoleErrorSpy: ReturnType<typeof spyOn>;
  let originalChalkLevel: typeof chalk.level;

  beforeEach(() => {
    // Désactiver les couleurs pour des tests reproductibles
    originalChalkLevel = chalk.level;
    chalk.level = 0;

    // Réinitialiser le niveau de log
    Logger.setLevel(LogLevel.DEBUG);
  });

  afterEach(() => {
    if (consoleLogSpy) consoleLogSpy.mockRestore();
    if (consoleWarnSpy) consoleWarnSpy.mockRestore();
    if (consoleErrorSpy) consoleErrorSpy.mockRestore();

    chalk.level = originalChalkLevel;
  });

  /**
   * Capture la sortie console et la normalise
   */
  const captureOutput = (fn: () => void): string => {
    const output: string[] = [];

    consoleLogSpy = spyOn(console, 'log').mockImplementation((...args) => {
      output.push(args.map((arg) => String(arg)).join(' '));
    });

    consoleWarnSpy = spyOn(console, 'warn').mockImplementation((...args) => {
      output.push(args.map((arg) => String(arg)).join(' '));
    });

    consoleErrorSpy = spyOn(console, 'error').mockImplementation((...args) => {
      output.push(args.map((arg) => String(arg)).join(' '));
    });

    fn();

    return output.join('\n');
  };

  test('debug: should display debug message at DEBUG level', () => {
    Logger.setLevel(LogLevel.DEBUG);

    const output = captureOutput(() => {
      Logger.debug('Debug message');
      Logger.debug('Debug with args', { key: 'value' });
    });

    expect(output).toMatchSnapshot();
  });

  test('debug: should not display at INFO level or higher', () => {
    Logger.setLevel(LogLevel.INFO);

    const output = captureOutput(() => {
      Logger.debug('This should not appear');
    });

    expect(output).toBe('');
    expect({ debugSuppressed: output === '' }).toMatchSnapshot();
  });

  test('info: should display info message', () => {
    const output = captureOutput(() => {
      Logger.info('Info message');
      Logger.info('Info with data', { count: 42 });
    });

    expect(output).toMatchSnapshot();
  });

  test('log: should display plain log message', () => {
    const output = captureOutput(() => {
      Logger.log('Plain log message');
      Logger.log('Log with args', 'arg1', 'arg2');
    });

    expect(output).toMatchSnapshot();
  });

  test('success: should display success message', () => {
    const output = captureOutput(() => {
      Logger.success('Operation successful');
      Logger.success('Success with result', { result: 'OK' });
    });

    expect(output).toMatchSnapshot();
  });

  test('warn: should display warning message', () => {
    const output = captureOutput(() => {
      Logger.warn('Warning message');
      Logger.warn('Warning with details', { issue: 'something' });
    });

    expect(output).toMatchSnapshot();
  });

  test('error: should display error message', () => {
    const output = captureOutput(() => {
      Logger.error('Error message');
      Logger.error('Error with details', { code: 500 });
    });

    expect(output).toMatchSnapshot();
  });

  test('config: should display config message', () => {
    const output = captureOutput(() => {
      Logger.config('Config loaded');
      Logger.config('Config path', '/path/to/config');
    });

    expect(output).toMatchSnapshot();
  });

  test('files: should display files message', () => {
    const output = captureOutput(() => {
      Logger.files('Processing file.ts');
      Logger.files('Found files', 10);
    });

    expect(output).toMatchSnapshot();
  });

  test('cleanup: should display cleanup message', () => {
    const output = captureOutput(() => {
      Logger.cleanup('Cleaning up temp files');
    });

    expect(output).toMatchSnapshot();
  });

  test('settings: should display settings message', () => {
    const output = captureOutput(() => {
      Logger.settings('Settings updated');
      Logger.settings('Setting value', 'key=value');
    });

    expect(output).toMatchSnapshot();
  });

  test('analysis: should display analysis message', () => {
    const output = captureOutput(() => {
      Logger.analysis('Analyzing code');
      Logger.analysis('Found issues', 5);
    });

    expect(output).toMatchSnapshot();
  });

  test('examples: should display examples message', () => {
    const output = captureOutput(() => {
      Logger.examples('Example command');
      Logger.examples('Usage', 'npm run test');
    });

    expect(output).toMatchSnapshot();
  });

  test('report: should display report message', () => {
    const output = captureOutput(() => {
      Logger.report('Generating report');
      Logger.report('Report stats', { total: 100 });
    });

    expect(output).toMatchSnapshot();
  });

  test('list: should display list item', () => {
    const output = captureOutput(() => {
      Logger.list('First item');
      Logger.list('Second item');
      Logger.list('Third item');
    });

    expect(output).toMatchSnapshot();
  });

  test('section: should display section header', () => {
    const output = captureOutput(() => {
      Logger.section('Section Title');
      Logger.section('Another Section', { meta: 'data' });
    });

    expect(output).toMatchSnapshot();
  });

  test('filePath: should display file path', () => {
    const output = captureOutput(() => {
      Logger.filePath('/path/to/file.ts');
      Logger.filePath('./relative/path.js');
    });

    expect(output).toMatchSnapshot();
  });

  test('label: should display label', () => {
    const output = captureOutput(() => {
      Logger.label('Label:');
      Logger.label('Status:', 'active');
    });

    expect(output).toMatchSnapshot();
  });

  test('colored: should display colored message', () => {
    const output = captureOutput(() => {
      Logger.colored('blue', 'Blue message');
      Logger.colored('red', 'Red message', { extra: 'data' });
    });

    expect(output).toMatchSnapshot();
  });

  test('newLine: should output blank line', () => {
    const output = captureOutput(() => {
      Logger.log('Before');
      Logger.newLine();
      Logger.log('After');
    });

    expect(output).toMatchSnapshot();
  });

  test('setLevel: should filter messages by log level', () => {
    // Test WARN level (should only show WARN and ERROR)
    Logger.setLevel(LogLevel.WARN);

    const output = captureOutput(() => {
      Logger.debug('Debug - should not appear');
      Logger.info('Info - should not appear');
      Logger.success('Success - should not appear');
      Logger.warn('Warning - should appear');
      Logger.error('Error - should appear');
    });

    expect(output).toMatchSnapshot();
  });

  test('all methods combined: should display all message types', () => {
    Logger.setLevel(LogLevel.DEBUG);

    const output = captureOutput(() => {
      Logger.debug('Debug message');
      Logger.info('Info message');
      Logger.success('Success message');
      Logger.warn('Warning message');
      Logger.error('Error message');
      Logger.config('Config message');
      Logger.files('Files message');
      Logger.settings('Settings message');
      Logger.analysis('Analysis message');
      Logger.report('Report message');
      Logger.list('List item');
      Logger.section('Section');
      Logger.filePath('/path');
      Logger.label('Label');
    });

    expect(output).toMatchSnapshot();
  });
});
