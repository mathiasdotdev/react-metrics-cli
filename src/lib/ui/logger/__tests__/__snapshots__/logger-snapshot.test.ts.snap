// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Logger - Snapshot Tests debug: should display debug message at DEBUG level 1`] = `
"🐛 Debug message
🐛 Debug with args [object Object]"
`;

exports[`Logger - Snapshot Tests debug: should not display at INFO level or higher 1`] = `
{
  "debugSuppressed": true,
}
`;

exports[`Logger - Snapshot Tests info: should display info message 1`] = `
"💡 Info message
💡 Info with data [object Object]"
`;

exports[`Logger - Snapshot Tests log: should display plain log message 1`] = `
"Plain log message
Log with args arg1 arg2"
`;

exports[`Logger - Snapshot Tests success: should display success message 1`] = `
"✅ Operation successful
✅ Success with result [object Object]"
`;

exports[`Logger - Snapshot Tests warn: should display warning message 1`] = `
"⚠️ Warning message
⚠️ Warning with details"
`;

exports[`Logger - Snapshot Tests error: should display error message 1`] = `
"❌ Error message
❌ Error with details"
`;

exports[`Logger - Snapshot Tests config: should display config message 1`] = `
"🔐 Config loaded
🔐 Config path /path/to/config"
`;

exports[`Logger - Snapshot Tests files: should display files message 1`] = `
"📁 Processing file.ts
📁 Found files 10"
`;

exports[`Logger - Snapshot Tests cleanup: should display cleanup message 1`] = `"🗑️ Cleaning up temp files"`;

exports[`Logger - Snapshot Tests settings: should display settings message 1`] = `
"⚙️ Settings updated
⚙️ Setting value key=value"
`;

exports[`Logger - Snapshot Tests analysis: should display analysis message 1`] = `
"🔍 Analyzing code
🔍 Found issues 5"
`;

exports[`Logger - Snapshot Tests examples: should display examples message 1`] = `
"✍️ Example command
✍️ Usage npm run test"
`;

exports[`Logger - Snapshot Tests report: should display report message 1`] = `
"📊 Generating report
📊 Report stats [object Object]"
`;

exports[`Logger - Snapshot Tests list: should display list item 1`] = `
"- First item
- Second item
- Third item"
`;

exports[`Logger - Snapshot Tests section: should display section header 1`] = `
"Section Title
Another Section [object Object]"
`;

exports[`Logger - Snapshot Tests filePath: should display file path 1`] = `
"/path/to/file.ts
./relative/path.js"
`;

exports[`Logger - Snapshot Tests label: should display label 1`] = `
"Label:
Status: active"
`;

exports[`Logger - Snapshot Tests colored: should display colored message 1`] = `
"Blue message
Red message [object Object]"
`;

exports[`Logger - Snapshot Tests newLine: should output blank line 1`] = `
"Before


After"
`;

exports[`Logger - Snapshot Tests setLevel: should filter messages by log level 1`] = `
"⚠️ Warning - should appear
❌ Error - should appear"
`;

exports[`Logger - Snapshot Tests all methods combined: should display all message types 1`] = `
"🐛 Debug message
💡 Info message
✅ Success message
⚠️ Warning message
❌ Error message
🔐 Config message
📁 Files message
⚙️ Settings message
🔍 Analysis message
📊 Report message
- List item
Section
/path
Label"
`;
